### PR TITLE
Add support for Amazon S3

### PR DESCRIPTION
**Pull request by daviddavis**
fixes #3900
https://pulp.plan.io/issues/3900

### CODE REVIEW

File: pulpcore/app/models/fields.py [hunk 1]
```diff
@@ -35,7 +35,11 @@ def pre_save(self, model_instance, add):
                                'Artifact storage. Files must be stored outside this location '
                                'prior to Artifact creation.'))
         file = super().pre_save(model_instance, add)
-        if file and file._committed and add:
+        if file and file._committed and add and str(file) != upload_to:
+            # the file needs to be moved into place
```
Comment (daviddavis) [inline pulpcore/app/models/fields.py]: This change is a bit questionable in my mind. The problem is that the following lines of code was being executed for uploads that were already in place (via upload_to). The read failed in the case where we're using S3 because of course the file is already stored in S3. So I'm adding a check to see if the file path is equal to upload_to.

Not sure if there's a better way to do this?

File: docs/installation/configuration.rst [hunk 2]
```diff
@@ -42,6 +42,25 @@ DATABASES
    `Django documentation on databases <https://docs.djangoproject.com/en/2
    .1/ref/settings/#databases>`_
 
+DEFAULT_FILE_STORAGE
+^^^^^^^^^^^^^^^^^^^^
+
+   By default, Pulp uses the local filesystem to store files. This can be configured though to
```
Comment (bmbouter) [inline docs/installation/configuration.rst]: This would be clearer if the first paragraph was all about the defaults and the second paragraph was all about S3. That way more backends could be added as their own paragraphs later too. What do you think?

File: pulpcore/app/models/fields.py [hunk 3]
```diff
@@ -35,7 +35,11 @@ def pre_save(self, model_instance, add):
...
             file._file = TemporaryDownloadedFile(open(file.name, 'rb'))
             file._committed = False
```
Comment (bmbouter) [inline pulpcore/app/models/fields.py]: It seems like this strangeness was already here, but we're calling super() twice with the same arguments. Should we reorder these statements so we are calling it only once?

File: docs/installation/configuration.rst [hunk 4]
```diff
@@ -42,6 +42,25 @@ DATABASES
...
+   Note that if you use S3 you should at a minimum also set ``AWS_ACCESS_KEY_ID``,
+   ``AWS_SECRET_ACCESS_KEY``, and ``AWS_STORAGE_BUCKET_NAME``. For more S3 configuratino options,
+   see the `django-storages documents <https://django-storages.readthedocs.io/en/latest/backends/amazon-S3.html>`_.
+
+MEDIA_ROOT
+^^^^^^^^^^
+
+   The location where Pulp will store files. By default this is `/var/lib/pulp/`.
+
+   If you're using S3, this will be the path within the bucket you're using.
```
Comment (bmbouter) [inline docs/installation/configuration.rst]: Does the user need to set this or is it set for them? If they need to take action that could be clearer.

File: docs/installation/configuration.rst [hunk 5]
```diff
@@ -42,6 +42,25 @@ DATABASES
    `Django documentation on databases <https://docs.djangoproject.com/en/2
    .1/ref/settings/#databases>`_
 
+DEFAULT_FILE_STORAGE
+^^^^^^^^^^^^^^^^^^^^
+
+   By default, Pulp uses the local filesystem to store files. This can be configured though to
+   alternatively use `Amazon S3 <https://aws.amazon.com/s3/>`_. To use S3, set
+   ``DEFAULT_FILE_STORAGE`` to ``pulpcore.app.models.storage.S3Storage``. The default option which
```
Comment (bmbouter) [inline docs/installation/configuration.rst]: What about adding a sentence that identifies that the user needs to `pip install`?

File: pulpcore/app/models/storage.py [hunk 6]
```diff
@@ -164,3 +164,14 @@ def get_tls_path(model, name):
         type(model).__name__,
         str(uuid4()),
         name)
+
+
+try:
+    # django-storages/boto3 are optional packages
+    from storages.backends.s3boto3 import S3Boto3Storage
+
+    class S3Storage(S3Boto3Storage):
```
Comment (bmbouter) [inline pulpcore/app/models/storage.py]: Why don't we have users use `storages.backends.s3boto3.S3Boto3Storage` directly?

File: pulpcore/content/handler.py [hunk 7]
```diff
@@ -286,3 +287,23 @@ def _save_content_artifact(self, download_result, content_artifact):
...
+        Handle response for file.
+
+        Depending on where the file storage (e.g. filesystem, S3, etc) this could be responding with
+        the file (filesystem) or a redirect (S3).
+
+        Args:
+            file (:class:`django.db.models.fields.files.FieldFile`: File to respond with
+
+        Raises:
+            :class:`aiohttp.web_exceptions.HTTPFound`: When we need to redirect to the file
```
Comment (bmbouter) [inline pulpcore/content/handler.py]: This also needs a `NotImplementedError` in its Raises section.

File: pulpcore/content/handler.py [hunk 8]
```diff
@@ -286,3 +287,23 @@ def _save_content_artifact(self, download_result, content_artifact):
...
+        """
```
Comment (bmbouter) [inline pulpcore/content/handler.py]: I believe it also needs a Returns section with `FileResponse` named in it.

File: pulpcore/app/models/fields.py (hunk 3 above)
Comment (daviddavis) [reply pulpcore/app/models/fields.py]: I tried looking at how to reorder this to only call super() once but I am not sure. This code is confusing and hard to follow. Would it make sense to open a separate refactor issue in redmine to try to rewrite this logic?

File: pulpcore/app/models/storage.py (hunk 6 above)
Comment (daviddavis) [reply pulpcore/app/models/storage.py]: We could but then we couldn't introduce custom logic or settings later on without having users change their configuration. If we're sure that we're not going to need to do that later on, then I am fine with making that change.

Comment (bmbouter) [reply pulpcore/app/models/storage.py]: I don't know if we would do this in the future, but when we have a reason to we can introduce it's usage as a configurable setting with that version. I think removing it would be good.

File: pulpcore/app/models/fields.py (hunk 3 above)
Comment (bmbouter) [reply pulpcore/app/models/fields.py]: I'm ok with an issue filed now to refactor this later.

File: docs/installation/storage.rst [hunk 9]
```diff
@@ -0,0 +1,45 @@
+Storage
```
Comment (bmbouter) [inline docs/installation/storage.rst]: These new docs are legit!

File: docs/installation/storage.rst [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+=======
+
+.. _storage:
+
+-----------
+
+  Pulp uses `django-storages <https://django-storages.readthedocs.io/>`_ to support multiple storage backends. If no backend is configured,
```
Comment (bmbouter) [inline docs/installation/storage.rst]: Can this be line wrapped at 100 chars?

File: pulpcore/app/models/fields.py (hunk 3 above)
Comment (daviddavis) [reply pulpcore/app/models/fields.py]: Ok, here we go: https://pulp.plan.io/issues/4382

Comment (bmbouter) [reply pulpcore/app/models/fields.py]: Thanks!

File: pulpcore/content/handler.py [hunk 11]
```diff
@@ -286,3 +287,27 @@ def _save_content_artifact(self, download_result, content_artifact):
...
+
+    def _handle_file_response(self, file):
+        """
+        Handle response for file.
+
+        Depending on where the file storage (e.g. filesystem, S3, etc) this could be responding with
+        the file (filesystem) or a redirect (S3).
+
+        Args:
+            file (:class:`django.db.models.fields.files.FieldFile`: File to respond with
```
Comment (bmbouter) [inline pulpcore/content/handler.py]: I missing a paren at the end here.
